### PR TITLE
Fix handling of `iconSize` and `iconColor` defaults for `ButtonStyleButton` subclasses.

### PR DESCRIPTION
**Pull request by gspencergoog**
## Description

Adds defaults that use tokens to define default `iconSize` and `iconColor` values. Previously, the Material 3 token values for button icon sizes and colors were not being used as defaults when the `ButtonStyleButton.defaultStyleOf` function returned the default values.

Adds tests to make sure appropriate `ButtonStyle` fields are populated when defaultStyle is called on buttons.

Updated documentation for `defaultStyleOf` to indicated that not _all_ fields need to be non-null, since some fields make sense to be null (e.g. `fixedSize`) because they would otherwise override the behavior of other fields in the same `ButtonStyle`.

## Tests
 - Added tests to make sure that the appropriate fields are non-null in the default button styles for each type of button.

### CODE REVIEW

File: packages/flutter/lib/src/material/button_style_button.dart
Comment (HansMuller) [inline packages/flutter/lib/src/material/button_style_button.dart:164]: The original text is a little confusing because "the style returned by [themeStyleOf]" sounds like developers are supposed to override themeStyleOf. What we're really saying is that you can override the style with a button-specific theme, like TextButtonTheme or ElevatedButtonTheme.

File: packages/flutter/lib/src/material/button_style_button.dart
Comment (HansMuller) [inline packages/flutter/lib/src/material/button_style_button.dart:173]: I don't think they all override other properties in the ButtonStyle although fixedSize does. Unfortunately a longer explanation might be needed.

**Review by HansMuller — COMMENTED** [review]
Just a few quick comments. This all looks good.

File: packages/flutter/lib/src/material/button_style_button.dart
Comment (gspencergoog) [reply packages/flutter/lib/src/material/button_style_button.dart:173]: Maybe I should just enumerate the properties that are allowed to be null and say why for each of them?

File: packages/flutter/lib/src/material/button_style_button.dart
Comment (gspencergoog) [reply packages/flutter/lib/src/material/button_style_button.dart]: I agree.  I'm not quite sure how to revise it, thouhgh.  How's this?

File: packages/flutter/lib/src/material/button_style_button.dart
Comment (HansMuller) [reply packages/flutter/lib/src/material/button_style_button.dart]: I think that would help, yes.

File: packages/flutter/lib/src/material/button_style_button.dart
Comment (HansMuller) [reply packages/flutter/lib/src/material/button_style_button.dart]: I think it might be better to just leave out the info about [themeStyle]. It's enough to say that you can override a button's style with its style parameter or with a button-specific theme like TextButtonTheme or ElevatedButtonTheme.

**Review by QuncCccccc — COMMENTED** [review]
Thanks so much for the fix! For MenuItemButton, we probably also need to add a default size for icon because from carbon it is still using 24 as the icon size:) This can fix the failed unit test in DropdownMenu.

This icon size token in menu is deprecated and we can use the List leading-icon token(it was 18 as well but with the latest token value, it is updated to 24).
Screenshot 2024-02-15 at 1 13 01 PM
Screenshot 2024-02-15 at 1 14 50 PM

**Comment by QuncCccccc** [comment]
Looks like ToggleButtons are built on top of TextButton, maybe we can also revert the icon size to 24 and icon color [here](https://github.com/flutter/flutter/blob/ea3d066237a5559d54ad778283e516e74af5e596/packages/flutter/lib/src/material/toggle_buttons.dart#L762) because they are not part of M3. Then we can fix the failures in toggle_button_test.dart.

**Comment by gspencergoog** [comment]
It also looks like there are Google golden file changes to fix.

**Comment by gspencergoog** [comment]
@QuncCccccc (and cc @Piinks ) I don't have time right now to land this: it requires changes in tests and code in Google code, but I do think it should be landed sometime.  I did make cl/650377998 and file b/352768033 as a start. Do you or someone on the Framework team have time to look at it?

**Comment by QuncCccccc** [comment]
I can put it in my queue and fix failures in G3 gradually:)Thanks for letting me know!

**Comment by gspencergoog** [comment]
Okay, the fixes in Google code are complete, so there should be few if any failures there now.

**Comment by gspencergoog** [comment]
Looks like the remaining issues in Google code are `iconSize` related.

**Comment by QuncCccccc** [comment]
Update: waiting for cl/655739008 to land:)

**Comment by QuncCccccc** [comment]
Once cl/658520909 is landed, I think we are good to go!

**Comment by gspencergoog** [comment]
@QuncCccccc Okay looks like the CL has landed.  I updated this PR to HEAD.  If the tests pass and you approve, then I can commit it!

File: packages/flutter/test/material/outlined_button_test.dart
Comment (QuncCccccc) [inline packages/flutter/test/material/outlined_button_test.dart:179]:
```suggestion
    const ColorScheme colorScheme = ColorScheme();
    final ThemeData theme = ThemeData.from(colorScheme: colorScheme);
```

File: packages/flutter/test/material/outlined_button_test.dart
Comment (QuncCccccc) [inline packages/flutter/test/material/outlined_button_test.dart:233]:
```suggestion
    const ColorScheme colorScheme = ColorScheme();
    final ThemeData theme = ThemeData.from(colorScheme: colorScheme);
```

File: packages/flutter/test/material/text_button_test.dart
Comment (QuncCccccc) [inline packages/flutter/test/material/text_button_test.dart:159]:
```suggestion
    const ColorScheme colorScheme = ColorScheme();
    final ThemeData theme = ThemeData.from(colorScheme: colorScheme);
```

File: packages/flutter/test/material/text_button_test.dart
Comment (QuncCccccc) [inline packages/flutter/test/material/text_button_test.dart:213]:
```suggestion
    const ColorScheme colorScheme = ColorScheme();
    final ThemeData theme = ThemeData.from(colorScheme: colorScheme);
```

File: packages/flutter/test/material/filled_button_test.dart
Comment (QuncCccccc) [inline packages/flutter/test/material/filled_button_test.dart:186]:
```suggestion
    const ColorScheme colorScheme = ColorScheme();
    final ThemeData theme = ThemeData.from(colorScheme: colorScheme);
```

File: packages/flutter/test/material/filled_button_test.dart
Comment (QuncCccccc) [inline packages/flutter/test/material/filled_button_test.dart:132]:
```suggestion
    const ColorScheme colorScheme = ColorScheme();
    final ThemeData theme = ThemeData.from(colorScheme: colorScheme);
```

File: packages/flutter/test/material/elevated_button_test.dart
Comment (QuncCccccc) [inline packages/flutter/test/material/elevated_button_test.dart:218]:
```suggestion
    const ColorScheme colorScheme = ColorScheme();
    final ThemeData theme = ThemeData.from(colorScheme: colorScheme);
```

File: packages/flutter/test/material/elevated_button_test.dart
Comment (QuncCccccc) [inline packages/flutter/test/material/elevated_button_test.dart:164]:
```suggestion
    const ColorScheme colorScheme = ColorScheme();
    final ThemeData theme = ThemeData.from(colorScheme: colorScheme);
```

**Review by QuncCccccc — APPROVED** [review]
Oh I thought I already approved this PR. LGTM! Thanks for the fix! Just left several same nits:)

**Comment by chrisbobbe** [comment]
After this change, if you pass a custom `foregroundColor` but you don't pass an `iconColor`, then the icon won't be your chosen custom color, right? Is this intended? If so, it seems like we should update the dartdoc on those fields:

```diff
-  /// The color for the button's [Text] and [Icon] widget descendants.
+  /// The color for the button's [Text] widget descendants.
   ///
   /// This color is typically used instead of the color of the [textStyle]. All
   /// of the components that compute defaults from [ButtonStyle] values
   /// compute a default [foregroundColor] and use that instead of the
   /// [textStyle]'s color.
   final MaterialStateProperty<Color?>? foregroundColor;
```

```diff
   /// The icon's color inside of the button.
-  ///
-  /// If this is null, the icon color will be [foregroundColor].
   final MaterialStateProperty<Color?>? iconColor;
```

**Pull request merged**
